samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java

```diff
@@ -230,17 +228,22 @@ public void waitForFinish() {
    * @param intStreams list of intermediate {@link StreamSpec}s
    * @throws Exception exception for latch timeout
    */
-  /* package private */ void createStreams(List<StreamSpec> intStreams) throws Exception {
+  /* package private */ void createStreams(List<StreamSpec> intStreams) throws TimeoutException {
     if (!intStreams.isEmpty()) {
+      CoordinationUtils coordinationUtils = createCoordinationUtils();
       if (coordinationUtils != null) {
-        Latch initLatch = coordinationUtils.getLatch(1, INIT_LATCH_ID);
-        LeaderElector leaderElector = coordinationUtils.getLeaderElector();
-        leaderElector.setLeaderElectorListener(() -> {
-            getStreamManager().createStreams(intStreams);
-            initLatch.countDown();
-          });
-        leaderElector.tryBecomeLeader();
-        initLatch.await(LATCH_TIMEOUT_MINUTES, TimeUnit.MINUTES);
+        try {
+          Latch initLatch = coordinationUtils.getLatch(1, INIT_LATCH_ID);
+          LeaderElector leaderElector = coordinationUtils.getLeaderElector();
+          leaderElector.setLeaderElectorListener(() -> {
+              getStreamManager().createStreams(intStreams);
+              initLatch.countDown();
+            });
+          leaderElector.tryBecomeLeader();
+          initLatch.await(LATCH_TIMEOUT_MINUTES, TimeUnit.MINUTES);
+        } finally {
```

Comment:
>> The problem you quoted exists even in current world. 

Beg to differ. 

Imagine what can happen in worst case. Let’s say there’re X processors as a part of standalone application deployment.  

Each processor joins after the predecessor processor in the group has acquired the lock(passed leader election) 
and created intermediate streams (depending upon time processor joining the group). 

X processors join one at a time (P1 triggers leader election and creates intermediate stream), then P2 triggers leader election(since P2 joined 
after the completion of Leader election only with P1 & 
intermediate stream creation by P1) acquires lock and creates intermediate stream, 
you can extend this to all the X processors.

With this processor joining order(say ‘X’ processor joining) after 
the change, we will have ‘X’ leader election & ‘X’ intermediate streams 
creation calls(Every new processor joining 
will have new leader election + intermediate stream creation call). 

With the coordinationUtils.close only at shutdown, `one` leader election and `one` intermediate streams creation.

This is my only concern(With the combination of our debounce, latch timeout 
this will be a big number for sufficiently large ‘X’). 

Since this incurs significant cost at worst case, what is the benefit of closing it after leader election here and not during shutdown??

>> zNode maintainence vs cleanup

This is already taken care of by background GC  which  cleans up old zkNodes related to an app(which were reminiscent of 
previous process group coordination, 
generated job models etc). There should be no 
manual/operational intervention needed.

>> With regards to release blocker 

I think the fix was to branch the zkRootPath into two(One for ZkJobCooordinator and other for ZkCoordinationService). 



Since we expect streams creation to be idempotent we don’t have to hold a lock just to do that(Idempotency at server side will take care of multiple req’s also at same time).

I’m not sure of the consequences of this (cost of multiple leader elections, multiple intermediate stream creation).

I don't think it's a good idea to change it very close to a release. 

Nevertheless, we could do it if there’s a consensus.